pulpcore/tests/unit/content/test_handler.py

```diff
@@ -67,3 +74,196 @@ def test_save_artifact_artifact_already_exists(c2, ra1, ra2, download_result_moc
+        ra = args[2]
+        assert isinstance(ra, RemoteArtifact)
+        assert ra.remote == remote
+        assert ra.url == f"{remote.url}/c123"
+        assert ra.content_artifact.relative_path == "c123"
+    finally:
+        await remote.adelete()
+        await distro.adelete()
+
+
+@pytest.mark.asyncio
+@pytest.mark.django_db
+async def test_pull_through_metadata_file(request123, monkeypatch):
+    """Requested path is for a metadata file. Don't save response."""
+    handler = Handler()
+    handler._stream_remote_artifact = AsyncMock()
+
+    remote = await create_remote()
+    monkeypatch.setattr(Remote, "get_remote_artifact_content_type", Mock(return_value=None))
+    distro = await create_distribution(remote)
+
+    try:
+        await handler._match_and_stream(f"{distro.base_path}/c123", request123)
+        remote.get_remote_artifact_content_type.assert_called_once_with("c123")
+        handler._stream_remote_artifact.assert_called_once()
+
+        _, kwargs = handler._stream_remote_artifact.call_args
+        assert kwargs.get("save_artifact", None) is False
+    finally:
+        await remote.adelete()
+        await distro.adelete()
+
+
+def test_pull_through_save_single_artifact_content(
+    remote123, request123, download_result_mock, monkeypatch
+):
+    """Ensure single-artifact content is properly saved on pull-through."""
+    handler = Handler()
+    remote123.get_remote_artifact_content_type = Mock(return_value=Content)
+    content_init_mock = Mock(return_value=Content())
+    monkeypatch.setattr(Content, "init_from_artifact_and_relative_path", content_init_mock)
+    ca = ContentArtifact(relative_path="c123")
+    ra = RemoteArtifact(url=f"{remote123.url}/c123", remote=remote123, content_artifact=ca)
+
+    # Content is saved during handler._save_artifact
+    artifact = handler._save_artifact(download_result_mock, ra, request=request123)
+
+    remote123.get_remote_artifact_content_type.assert_called_once_with("c123")
+    content_init_mock.assert_called_once_with(artifact, "c123")
+
+    # Assert the CA and RA are properly saved
+    ca = artifact.content_memberships.first()
+    assert ca.content is not None
+    assert ca.relative_path == "c123"
+    ra = RemoteArtifact.objects.filter(
+        url=f"{remote123.url}/c123", remote=remote123, content_artifact=ca
+    ).first()
+    assert ra is not None
+
+
+def test_pull_through_save_multi_artifact_content(
+    remote123, request123, download_result_mock, monkeypatch, tmp_path
+):
+    """Ensure multi-artifact content is properly saved on pull-through."""
+    handler = Handler()
+    remote123.get_remote_artifact_content_type = Mock(return_value=Content)
+
+    tmp_file = tmp_path / str(uuid.uuid4())
+    tmp_file.write_text(str(tmp_file))
+    artifact123 = Artifact.init_and_validate(str(tmp_file))
+    artifact123.save()
+
+    def content_init(art, path):
+        return Content(), {"c123": artifact123, "abc123": art}
```

Comment:
```suggestion
        return Content(), {path: artifact123, "ab" + path: art}
```
Maybe I'm misunderstanding the logic, but i think it should _use_ the path.